File: backend/logger.py
# Backend logging utility for debugging API communication
import json
import os
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, Optional

class BackendLogger:
    def __init__(self, log_file: str = "communication.log"):
        self.log_file = Path(log_file)
        self.logs = []
        self.max_logs = 1000
        
    def log(self, level: str, source: str, message: str, data: Optional[Dict[str, Any]] = None):
        """Log an entry with timestamp"""
        entry = {
            "timestamp": datetime.utcnow().isoformat(),
            "level": level,
            "source": source,
            "message": message,
            "data": data or {}
        }
        
        # Add to memory
        self.logs.append(entry)
        
        # Keep only the last max_logs entries
        if len(self.logs) > self.max_logs:
            self.logs = self.logs[-self.max_logs:]
        
        # Write to file
        try:
            with open(self.log_file, "a", encoding="utf-8") as f:
                f.write(json.dumps(entry) + "\n")
        except Exception as e:
            print(f"Failed to write to log file: {e}")
        
        # Also print to console
        print(f"[{entry['timestamp']}] {level} [{source}] {message}")
        if data:
            print(f"  Data: {json.dumps(data, indent=2)}")
    
    def info(self, source: str, message: str, data: Optional[Dict[str, Any]] = None):
        self.log("INFO", source, message, data)
    
    def error(self, source: str, message: str, data: Optional[Dict[str, Any]] = None):
        self.log("ERROR", source, message, data)
    
    def warn(self, source: str, message: str, data: Optional[Dict[str, Any]] = None):
        self.log("WARN", source, message, data)
    
    def debug(self, source: str, message: str, data: Optional[Dict[str, Any]] = None):
        self.log("DEBUG", source, message, data)
    
    def get_logs(self) -> list:
        """Get all logs from memory"""
        return self.logs.copy()
    
    def export_logs(self) -> str:
        """Export logs as JSON string"""
        return json.dumps(self.logs, indent=2)
    
    def clear_logs(self):
        """Clear logs from memory and file"""
        self.logs = []
        try:
            if self.log_file.exists():
                self.log_file.unlink()
        except Exception as e:
            print(f"Failed to clear log file: {e}")

# Global logger instance
logger = BackendLogger()

File: frontend/dwas-visualizer-main/src/utils/logger.ts
// Frontend logging utility for debugging API communication
interface LogEntry {
  timestamp: string
  level: 'INFO' | 'ERROR' | 'WARN' | 'DEBUG'
  source: 'FRONTEND' | 'API'
  message: string
  data?: any
}

class Logger {
  private logs: LogEntry[] = []
  private maxLogs = 1000

  log(level: LogEntry['level'], source: LogEntry['source'], message: string, data?: any) {
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      source,
      message,
      data
    }
    
    this.logs.push(entry)
    
    // Keep only the last maxLogs entries
    if (this.logs.length > this.maxLogs) {
      this.logs = this.logs.slice(-this.maxLogs)
    }
    
    // Also log to console
    const consoleMessage = `[${entry.timestamp}] ${entry.level} [${entry.source}] ${entry.message}`
    if (data) {
      console.log(consoleMessage, data)
    } else {
      console.log(consoleMessage)
    }
  }

  info(source: LogEntry['source'], message: string, data?: any) {
    this.log('INFO', source, message, data)
  }

  error(source: LogEntry['source'], message: string, data?: any) {
    this.log('ERROR', source, message, data)
  }

  warn(source: LogEntry['source'], message: string, data?: any) {
    this.log('WARN', source, message, data)
  }

  debug(source: LogEntry['source'], message: string, data?: any) {
    this.log('DEBUG', source, message, data)
  }

  getLogs(): LogEntry[] {
    return [...this.logs]
  }

  exportLogs(): string {
    return JSON.stringify(this.logs, null, 2)
  }

  downloadLogs() {
    const logs = this.exportLogs()
    const blob = new Blob([logs], { type: 'application/json' })
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `dwas-frontend-logs-${new Date().toISOString().split('T')[0]}.json`
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(a)
    URL.revokeObjectURL(url)
  }

  clearLogs() {
    this.logs = []
  }
}

export const logger = new Logger()

// Log all fetch requests
const originalFetch = window.fetch
window.fetch = async (...args) => {
  const url = args[0] as string
  const options = args[1] as RequestInit
  
  logger.info('API', `Making ${options?.method || 'GET'} request to ${url}`, {
    url,
    method: options?.method || 'GET',
    headers: options?.headers,
    body: options?.body
  })
  
  try {
    const response = await originalFetch(...args)
    
    logger.info('API', `Response from ${url}: ${response.status} ${response.statusText}`, {
      url,
      status: response.status,
      statusText: response.statusText,
      headers: Object.fromEntries(response.headers.entries())
    })
    
    return response
  } catch (error) {
    logger.error('API', `Request failed for ${url}`, {
      url,
      error: error instanceof Error ? error.message : String(error)
    })
    throw error
  }
}
